### ISSUE TITLE
Problem density matrix parsing for ROHF calculation

### ISSUE THREAD

**Issue by rbenda** [cclib/cclib#836]
Good evening everyone,
I am writing here following Adam Tenderholt advice, whom I contacted, about a possible problem of cclib concerning the parsing of alpha and beta density matrices (construction of the density matrix with the density.py code in the Method class), for the case of Restricted Open-Shell calculations such as ROHF.

For Unrestricted calculations giving rise to different alpha and beta molecular orbitals, the parsing of 'mocoeffs', 'homos' attributes is correctly done by cclib and the calculated density matrix (from alpha and beta density matrices) is perfectly correct.

However in the case of an ROHF calculation, cclib does not treat it as a 'unrestricted' calculation (I saw also the discution on this forum about ROHF, and the problem seems not to be fully solved if I understand well). However, there are also alpha and beta density matrices, as in the UHF cases (as you can see from the Gaussian output attached, calculation on a Copper 2+ ion). The mocoeffs attribute parsed by cclib in the case of an ROHF is correct, but the 'homos' is not (should be [12,13] instead of [13] here) and  the density matrix computed using density.py code of the 'Method' class of cclib is not correct. (which is the original problem I encountered)
Indeed, only one density matrix is given i.e. len(d.density)==1 (and it is not the sum of alpha and beta density matrices), but there should be two (called alpha and beta density matrices also ; and accounting than more electrons have alpha spin than beta, due to open shells).

Thus when computing the total charge from this density matrix computed with cclib, the result is not correct . (I find +28 instead of +27 for the electronic charge [number of electrons] of Cu2+ ; as if all shells are counted as doubly occupied ; yet in ROHF or RO-DFT calculations there are singly occupied orbitals ...).

As an example I am sending you this file of a ROHF Gaussian calculation on Cu(2+) ions (with one single alpha electron). But the problem is the same for ROHF on Ni2+, Co2+, etc. Singly occupied orbitals are considered as doubly occupied.

Additionally I would like to thank you  for all your work on cclib ; which is an outstanding and very convenient library to parse QM output files ! :)
Best regards,
Robert Benda
[CU2_ion_alone_ROHF_6-31Gd_no_PCM_pop_full_iop_3_33.log](https://github.com/cclib/cclib/files/4452452/CU2_ion_alone_ROHF_6-31Gd_no_PCM_pop_full_iop_3_33.log)

**Comment by rbenda** [comment]
Actually, if I am not mistaken, there is the same problem as the one underlined above for a Gaussian output file, of wrong density matrix parsed for GAMESS ROHF output files (maybe this is a general problem of wrong density matrix for all the ROHF QM calculations parsed by cclib).

 An example of GAMESS .log file for a Fe3+ ion is attached. The total charge recovered from the density matrix parsed with Density() cclib.method function is 28 (=2 times the number of alpha electrons [14]), instead of 23 (14 alpha electrons, 9 beta electrons), as the Density function wrongly treats all orbitals as doubly occupied (as in a 'restricted' calculation, although the 'inrestricted' flag should be also activated in the 'density.py' code)

[Fe3+_ROHF_6-31Gd.log](https://github.com/cclib/cclib/files/4503809/Fe3%2B_ROHF_6-31Gd.log)

Thank you very much,
Best regards
Robert Benda

**Comment by berquist** [comment]
> maybe this is a general problem of wrong density matrix for all the ROHF QM calculations parsed by cclib

Yes, this is a problem with almost all of our parsers, and exists at both the parser level and in the `Density` method. See https://github.com/cclib/cclib/issues/301 for comments on dealing with ROHF calculations at the parser level.

**Comment by rbenda** [comment]
Dear all,
following this problem of density matrix parsing for ROHF calculations, I think I ended up with a code which could adapt the density.py code of cclib (in the method repository) :

nb_basis_functions : to be defined thanks to cclib attribute

density_matrix=np.zeros([nb_basis_functions,nb_basis_functions]),np.zeros([nb_basis_functions,nb_basis_functions])]
    
total_nb_electrons=np.sum(data.atomnos)-data.charge
#Number of doubly occupied orbitals :
nb_occupied_MO=int(total_nb_electrons/2)

#Number of singly occupied orbitals :
Ns=total_nb_electrons-2*nb_occupied_MO

for i in range(nb_basis_functions):
for j in range(nb_basis_functions):
{
#*** Sum on DOUBLY OCCUPIED molecular orbitals
for k in range(nb_occupied_MO):
    density_matrix[0][i][j]+=2*data.mocoeffs[0][k][i]*data.mocoeffs[0][k][j]

#*** Sum on SINGLY OCCUPIED molecular orbitals (=> no factor 2)
for k in range(nb_occupied_MO,nb_occupied_MO+Ns+1):     
    density_matrix[1][i][j]+=data.mocoeffs[0][k][i]*data.mocoeffs[0][k][j]    
}
###
#Checking the total charge :
Q_tot=0
    
for mu in range(nb_basis_functions):
for nu in range(nb_basis_functions):
 {      
Q_tot+=density_matrix[0][mu][nu]*overlap_matrix[mu][nu]
Q_tot+=0.5*density_matrix[1][mu][nu]*overlap_matrix[mu][nu]
}

We find the correct total number of electrons 'Q_tot' for open shell systems, by  this technique, taking of course the correct overlap matrix (beware of some existing problems with aooverlaps attribute : [https://github.com/cclib/cclib/issues/1009](url) )

Thanks,
Robert Benda
